### PR TITLE
Post Creation Exercise Information Usecase

### PR DESCRIPTION
**Pull request by Ludosss**
Creation of usecase specific classes for Exercise Information usecase (User receiving exercise information for use in post creation)

### CODE REVIEW

File: src/main/java/ca/utoronto/fitbook/application/service/ViewExerciseInfoService.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
+package ca.utoronto.fitbook.application.service;
+
+
```
Comment (modaserMoj) [inline src/main/java/ca/utoronto/fitbook/application/service/ViewExerciseInfoService.java:3]: Remove space line 3

File: src/main/java/ca/utoronto/fitbook/application/port/in/LoadAllExercisesPort.java [hunk 2]
```diff
@@ -0,0 +1,9 @@
+package ca.utoronto.fitbook.application.port.in;
+
+import ca.utoronto.fitbook.entity.Exercise;
+
+import java.util.ArrayList;
```
Comment (modaserMoj) [inline src/main/java/ca/utoronto/fitbook/application/port/in/LoadAllExercisesPort.java:5]: remove space line 4

File: src/main/java/ca/utoronto/fitbook/adapter/web/ViewExerciseInfoController.java [hunk 3]
```diff
@@ -0,0 +1,33 @@
+package ca.utoronto.fitbook.adapter.web;
+
+import ca.utoronto.fitbook.adapter.persistence.UnauthorizedUserException;
+import ca.utoronto.fitbook.application.port.in.ViewExerciseInfoUseCase;
+import ca.utoronto.fitbook.application.port.out.response.ViewExerciseInfoResponse;
+import lombok.RequiredArgsConstructor;
+import org.springframework.stereotype.Controller;
+import org.springframework.ui.Model;
+import org.springframework.web.bind.annotation.GetMapping;
+
```
Comment (modaserMoj) [inline src/main/java/ca/utoronto/fitbook/adapter/web/ViewExerciseInfoController.java:10]: Remove space line 10

File: src/main/java/ca/utoronto/fitbook/adapter/persistence/localmemory/ExerciseLocalMemoryRepository.java [hunk 4]
```diff
@@ -2,6 +2,7 @@
 
 import ca.utoronto.fitbook.adapter.persistence.GenericRepository;
 import ca.utoronto.fitbook.application.exceptions.EntityNotFoundException;
+import ca.utoronto.fitbook.application.port.in.LoadAllExercisesPort;
 import ca.utoronto.fitbook.application.port.in.LoadExerciseListPort;
 import ca.utoronto.fitbook.entity.Exercise;
```
Comment (modaserMoj) [inline src/main/java/ca/utoronto/fitbook/adapter/persistence/localmemory/ExerciseLocalMemoryRepository.java:8]: remove space line 8

File: src/main/java/ca/utoronto/fitbook/adapter/web/ViewExerciseInfoController.java [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+    @GetMapping(path = "/post")
+    String viewInfo(Model model, HttpSession session){
+
+        // Get session userId, throw error if null
+        String userId = (String) session.getAttribute("userId");
+        if (userId == null)
+            throw new UnauthorizedUserException();
+
+        // Create output data using service method
+        ViewExerciseInfoResponse outputData = viewExerciseInfoUseCase.viewInfo();
```
Comment (modaserMoj) [inline src/main/java/ca/utoronto/fitbook/adapter/web/ViewExerciseInfoController.java:28]: tab

File: src/main/java/ca/utoronto/fitbook/application/service/ViewExerciseInfoService.java [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+@Service
+@RequiredArgsConstructor
+public class ViewExerciseInfoService implements ViewExerciseInfoUseCase {
+
+    private final LoadAllExercisesPort loadAllExercisesPort;
+
+    @Override
+    public ViewExerciseInfoResponse viewInfo() {
+        ArrayList<Exercise> exerciseList = loadAllExercisesPort.loadAllExercises();
```
Comment (modaserMoj) [inline src/main/java/ca/utoronto/fitbook/application/service/ViewExerciseInfoService.java:21]: tab

File: src/main/java/ca/utoronto/fitbook/adapter/persistence/firebase/ExerciseFirebaseRepository.java [hunk 7]
```diff
@@ -89,4 +87,40 @@ public List<Exercise> loadExerciseList(List<String> exerciseIds) throws EntityNo
...
+        return exercises;
+    }
+
+    /**
+     * Helper function to retrieve exercise object from document snapshot
+     * @param document in database pointing to specific exercise
+     * @return exercise located within document
+     * @throws ClassNotFoundException if the exerciseClassName cannot be located
+     */
+    public Exercise getExerciseHelper(DocumentSnapshot document) throws ClassNotFoundException {
```
Comment (modaserMoj) [inline src/main/java/ca/utoronto/fitbook/adapter/persistence/firebase/ExerciseFirebaseRepository.java:116]: tab to align code

File: src/main/java/ca/utoronto/fitbook/application/service/ViewExerciseInfoService.java (hunk 1 above)
Comment (Ludosss) [reply src/main/java/ca/utoronto/fitbook/application/service/ViewExerciseInfoService.java:3]: done

File: src/main/java/ca/utoronto/fitbook/application/port/in/LoadAllExercisesPort.java (hunk 2 above)
Comment (Ludosss) [reply src/main/java/ca/utoronto/fitbook/application/port/in/LoadAllExercisesPort.java:5]: done

File: src/main/java/ca/utoronto/fitbook/adapter/web/ViewExerciseInfoController.java (hunk 3 above)
Comment (Ludosss) [reply src/main/java/ca/utoronto/fitbook/adapter/web/ViewExerciseInfoController.java:10]: done

File: src/main/java/ca/utoronto/fitbook/adapter/persistence/localmemory/ExerciseLocalMemoryRepository.java (hunk 4 above)
Comment (Ludosss) [reply src/main/java/ca/utoronto/fitbook/adapter/persistence/localmemory/ExerciseLocalMemoryRepository.java:8]: done

**Review by modaserMoj — DISMISSED** [review]
seems alright follows the required guidelines

File: src/main/java/ca/utoronto/fitbook/application/service/ViewExerciseInfoService.java [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+
```
Comment (youssefsoli) [inline src/main/java/ca/utoronto/fitbook/application/service/ViewExerciseInfoService.java:22]: Remove this newline.

File: src/main/java/ca/utoronto/fitbook/application/port/out/response/ViewExerciseInfoResponse.java [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+
+import ca.utoronto.fitbook.entity.Exercise;
+import lombok.NonNull;
+import lombok.Value;
+
+import java.util.ArrayList;
+
+@Value
+public class ViewExerciseInfoResponse {
+
```
Comment (youssefsoli) [inline src/main/java/ca/utoronto/fitbook/application/port/out/response/ViewExerciseInfoResponse.java:11]: Extraneous newline.

File: src/main/java/ca/utoronto/fitbook/adapter/web/ViewExerciseInfoController.java [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+        model.addAttribute("exerList", outputData.getExercises());
+
+        return "InfoPage";
```
Comment (youssefsoli) [inline src/main/java/ca/utoronto/fitbook/adapter/web/ViewExerciseInfoController.java:31]: The template files should not be capitalized.

File: src/main/java/ca/utoronto/fitbook/adapter/web/ViewExerciseInfoController.java [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+    String viewInfo(Model model, HttpSession session){
+
+        // Get session userId, throw error if null
+        String userId = (String) session.getAttribute("userId");
+        if (userId == null)
+            throw new UnauthorizedUserException();
+
+        // Create output data using service method
+        ViewExerciseInfoResponse outputData = viewExerciseInfoUseCase.viewInfo();
+        model.addAttribute("exerList", outputData.getExercises());
```
Comment (youssefsoli) [inline src/main/java/ca/utoronto/fitbook/adapter/web/ViewExerciseInfoController.java:29]:
```suggestion
        model.addAttribute("exerciseList", outputData.getExercises());
```

File: src/main/java/ca/utoronto/fitbook/adapter/persistence/firebase/ExerciseFirebaseRepository.java [hunk 12]
```diff
@@ -102,4 +100,40 @@ public List<Exercise> loadExerciseList(List<String> exerciseIds) throws EntityNo
...
+     * @return A list of all exercises in the database
+     */
+    @Override
+    public ArrayList<Exercise> loadAllExercises() {
+        Iterable<DocumentReference> future = firestore.collection(COLLECTION_NAME).listDocuments();
+        ArrayList<Exercise> exercises = new ArrayList<>();
+        for (DocumentReference reference : future) {
+            ApiFuture<DocumentSnapshot> snapshot = reference.get();
+            try {
+                exercises.add(getExerciseHelper(snapshot.get()));
```
Comment (youssefsoli) [inline src/main/java/ca/utoronto/fitbook/adapter/persistence/firebase/ExerciseFirebaseRepository.java:115]: You should be able to utilize the `.toObject` method if you `.get()` the future snapshot for a DocumentSnapshot.

File: src/main/java/ca/utoronto/fitbook/adapter/persistence/firebase/ExerciseFirebaseRepository.java [hunk 13]
```diff
@@ -102,4 +100,40 @@ public List<Exercise> loadExerciseList(List<String> exerciseIds) throws EntityNo
         }
         return exerciseList;
     }
+
+    /**
+     *
```
Comment (youssefsoli) [inline src/main/java/ca/utoronto/fitbook/adapter/persistence/firebase/ExerciseFirebaseRepository.java:105]: Extra new line here.

File: src/main/java/ca/utoronto/fitbook/adapter/persistence/localmemory/ExerciseLocalMemoryRepository.java [hunk 14]
```diff
@@ -54,5 +58,14 @@ public List<Exercise> loadExerciseList(List<String> exerciseIds) throws EntityNo
         }
         return exerciseList;
     }
+
+    /**
+     *
```
Comment (youssefsoli) [inline src/main/java/ca/utoronto/fitbook/adapter/persistence/localmemory/ExerciseLocalMemoryRepository.java:63]: Extra newline.

File: src/main/java/ca/utoronto/fitbook/application/port/out/response/ViewExerciseInfoResponse.java [hunk 15]
```diff
@@ -0,0 +1,14 @@
...
+    @NonNull
+    ArrayList<Exercise> exercises;
```
Comment (youssefsoli) [inline src/main/java/ca/utoronto/fitbook/application/port/out/response/ViewExerciseInfoResponse.java:13]: I think we should return two lists, one for temporal exercises and one for repetitive exercises.

Comment (youssefsoli) [reply src/main/java/ca/utoronto/fitbook/application/port/out/response/ViewExerciseInfoResponse.java:13]: We can achieve this by checking the type of exercises coming out of the repository.

**Review by youssefsoli — DISMISSED** [review]

**Review by jannater — DISMISSED** [review]
Looks good after improvising.

**Review by modaserMoj — APPROVED** [review]
Code has been modified given the feedback. Great work!

**Review by jpablo2002 — APPROVED** [review]
Good completion of a core functionality for the user in the front end  to create posts following Clean Architecture

**Review by feiwun — APPROVED** [review]
code implements intended purposes

**Pull request merged**